pkg/mapper/dynamicfile/dynamicfile.go

```diff
@@ -0,0 +1,262 @@
+}
+
+type ErrParsingMap struct {
+	errors []error
+}
+
+func (err ErrParsingMap) Error() string {
+	return fmt.Sprintf("error parsing dynamic file: %v", err.errors)
+}
+
+func waitUntilFileAvailable(filename string) error {
+	waitFileTimer := time.NewTimer(1 * time.Second)
+	for {
+		<-waitFileTimer.C
+		_, err := os.Stat(filename)
+		if err != nil {
+			if os.IsNotExist(err) {
+				waitFileTimer.Reset(1 * time.Second)
+				continue
+			} else {
+				return err
+			}
+		}
+		break
+	}
+	return nil
+}
+
+func NewDynamicFileMapStore(filename string) (*DynamicFileMapStore, error) {
+	ms := DynamicFileMapStore{}
+	ms.filename = filename
+	return &ms, nil
+}
+
+func (m *DynamicFileMapStore) startLoadDynamicFile(stopCh <-chan struct{}) {
+	go func() {
+		//wait for the file
+		err := waitUntilFileAvailable(m.filename)
+		if err != nil {
+			logrus.Errorf("startLoadDynamicFile: failed to wait till dynamic file available %v", err)
+		}
+
+		//load the initial file content into memory
+		userMappings, roleMappings, awsAccounts, err := ParseMap(m.filename)
+		if err != nil {
+			logrus.Errorf("startLoadDynamicFile: There was an error parsing the dynamic file: %+v. Map is not updated. Please correct dynamic file", err)
+		} else {
+			m.saveMap(userMappings, roleMappings, awsAccounts)
+		}
+		//start to watch the file change
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			logrus.Errorf("startLoadDynamicFile: failed when call fsnotify.NewWatcher, %+v", err)
+		}
+		err = watcher.Add(m.filename)
+		if err != nil {
+			logrus.Errorf("startLoadDynamicFile: could not add file to watcher %v", err)
+		}
+
+		defer watcher.Close()
+		for {
+			select {
+			case <-stopCh:
+				return
+			case event := <-watcher.Events:
+				switch {
+				case event.Op&fsnotify.Write == fsnotify.Write, event.Op&fsnotify.Create == fsnotify.Create:
+					//reload the access entry file
+					logrus.Info("startLoadDynamicFile: got WRITE/CREATE event reload it the memory")
+					userMappings, roleMappings, awsAccounts, err := ParseMap(m.filename)
+					if err != nil {
+						logrus.Errorf("startLoadDynamicFile: There was an error parsing the dynamic file during handling WRITE event: %+v. Map is not updated. Please correct dynamic file", err)
+					} else {
+						m.saveMap(userMappings, roleMappings, awsAccounts)
+					}
+				case event.Op&fsnotify.Rename == fsnotify.Rename, event.Op&fsnotify.Remove == fsnotify.Remove:
+					logrus.Info("startLoadDynamicFile: got RENAME/REMOVE event")
+					//test if the "REMOVE" is triggered by vi or cp cmd
+					_, err := os.Stat(m.filename)
+					if os.IsNotExist(err) {
+						//the "REMOVE" event is  not triggered by vi or cp cmd
+						//reset memory
+						userMappings := make([]config.UserMapping, 0)
+						roleMappings := make([]config.RoleMapping, 0)
+						awsAccounts := make([]string, 0)
+						m.saveMap(userMappings, roleMappings, awsAccounts)
+					}
+					//wait for the file
+					err = waitUntilFileAvailable(m.filename)
```

Comment:
If using my suggestion above, this should end the goroutine and it should be restarted by something like `go wait.Until(watchDynamicFile, time.Second, ctx.Done())`.  This will remove the need for duplicate calls to `waitUntilFileAvailable()`.

Reply:
I had a comment about a refactor as well but looks like it wasn't saved. Another idea I had for a refactor was:
```
for {
  select {
    case <-stopCh:
    case fsnotifyevent: // any event
       // add mappings if file updated
       // clear mappings if file removed
    default:
       setupFileIfAvailable
       // this bit can do the job of waitUntilFileAvailable
  }
}
```

This would remove the fencepost problem of settting up a file before entering the forloop, and also consolidate some of the duplicate code that's in the select statement right now. 

Just an option :)